### ISSUE TITLE
Fail2ban doesn't notice configs in jail.d

### ISSUE THREAD

**Issue by swelljoe** [webmin/webmin#599]
Fail2ban supports configuration files in /etc/fail2ban/jail.d, which the Webmin module seems to ignore.

It is used in distribution packages to setup the default action. In the case of CentOS 7, it changes the default action to firewallcmd-ipset, but Webmin still says it is iptables-multiport. I'm using it to add that rule on Debian/Ubuntu, as well, even though they have their own defaults

As far as I can tell from my testing, the files in jail.d are processed after .conf files but before .local files. So, jail.d/whatever.conf will override jail.conf, but will be overridden by jail.local.

You can see what configuration fail2ban has ended up with by using the fail2ban client to query it. e.g.

`fail2ban-client get sshd actions`

After making changes, you can use the reload `fail2ban-client reload`.

The /etc/fail2ban/jail.d/00-firewalld.conf I'm generating for Ubuntu/Debian is copied from CentOS 7 and looks like this (but anything that can appear in jail.conf can appear in these files...I guess they execute in alphabetic order, due to CentOS 7 prepending 00):

```
[DEFAULT]
banaction = firewallcmd-ipset
```

**Comment by chris001** [comment]
Yes, fail2ban integration is incomplete.  It's often running invisibly, while the virtualmin interface isn't reporting any of its detection or blocking activity, or accurately letting you modify its settings to block new patterns of attack. Result, many hacking attempts take place, while virtualmin shows no notice to the admin. This needs to be completely implemented.

**Comment by swelljoe** [comment]
This ticket isn't about Virtualmin integration of fail2ban. It's about a specific issue with the Webmin fail2ban module, where it isn't reading in some config files that are in common use.

We'll get Virtualmin more aware of fail2ban once it's working in the installer, but that's a separate issue. Let's make this trip one actionable step at a time! ;-)

**Comment by jcameron** [comment]
That's odd, because Webmin should already be reading files from the `jail.d` directory as long as they have a `.conf` extension.

**Comment by jcameron** [comment]
The code that does this : https://github.com/webmin/webmin/blob/master/fail2ban/fail2ban-lib.pl#L88

**Comment by swelljoe** [comment]
Huh...so, it should come before .local, in order to match the behavior of fail2ban itself.

But, let me test and see if it's just the order that's wrong, or if there's another issue.

**Comment by swelljoe** [comment]
Yeah, so it doesn't seem to be picking up the changed banaction. jail.local works, but anything in jail.d is unable to override anything in jail.conf (which it does in fail2ban itself).

The way it ought to work is: parse jail.conf first, then read jail.d/\*.conf overriding anything already defined in jail.conf, and then finally jail.local can override any of those. So, it looks like local is special-cased inside of the parse function, so I think jail.d needs that same treatment.

**Comment by gnadelwartz** [comment]
we sould parse like  faik2ban do, but until now I was also assuming files in conf.d can not overwrite global config. because if an adittional installed program place a file there which overwrite e.g. default action with an errornous one... this will be very bad.

**Comment by swelljoe** [comment]
If a package places a file there and it doesn't work as expected, that's also very bad, too, right?

Global defaults can be overridden in either .local files on in jail.d files (with jail.d overriding jail.conf, and jail.local overriding both), and it's how it is used by packagers on CentOS, Debian, and Ubuntu. On CentOS/Fedora/RHEL it is used to make it easy to choose iptables or firewalld behavior just by installing a single package (fail2ban-firewalld), which just adds a file containing the config I listed above in jail.d.

Anyway, regardless of how any of us think it ought to work...the way it actually works is what I described in the comment above, so that's the thing we gotta do. If Jamie doesn't have time to tackle it, I'll take a look at it and see if I can figure it out.

**Comment by gnadelwartz** [comment]
iam withyou, it should parsed the way fail2ban does, i was only suprised about.

**Comment by gnadelwartz** [comment]
the problem seems to be in` parse_config_file`, because it does not overwrite already existing config, but internally checks for a `.local` version of the file and override only the values of the original given file.

**Comment by gnadelwartz** [comment]
after a more deep look, `parse_config_file` IMHO it does not know about any existing config. It simply reads in a new config and retrurns the values. `list_jails` simply push the returned config to the end of the existing config.

I see two possibilities:

1.  extend parsing of config array to use always he latest found entry
2.  add a function to override already existing config while read from file, like `parse_config_file`does it internally for .local file.

not tested example code for 2.  stolen from `parse_config_file`:

```
#sub parse_config_file_override(\@conf, $file)
sub parse_config_file_override
{
my ($conf,$file) = @_;

my @lrv = &parse_config_file($file);
my %lsects = map { $_->{'name'}, $_ } @lrv;
        foreach my $sect ($conf) {
                # For each section in the .conf file, put directives
                # from the .local file in first
                my $lsect = $lsects{$sect->{'name'}};
                next if (!$lsect);
                my $msect = { %$lsect };
                $msect->{'local'} = 1;
                $msect->{'origfile'} = $sect->{'file'};
                push(@{$msect->{'members'}}, @{$sect->{'members'}});
                $rv[&indexof($sect, @rv)] = $msect;
                }
        }
return $conf;
}

```

**Comment by swelljoe** [comment]
It seems like the simplest path is to remove the special case for .local (or anything else), and just parse the files in order (jail.conf, jail.d/, jail.local), while inserting the values as they're found...overwriting if they already exist. It seems like the current code is only allowing overriding of some things, but that's not how fail2ban is parsing it...any file can set any value, and when it is parsed is what decides whether it is the final value.

I've got a couple of other things to finish up today, but will hopefully be able to look at it later.

**Comment by jcameron** [comment]
So do the .local files completely override the .conf files, or do they just update individual directives?

**Comment by gnadelwartz** [comment]
it overrides individual existing directives 😉 and add new ones @ the end.

**Comment by swelljoe** [comment]
They override individual directives. Any directive defined in jail.conf can be overridden by jail.d/\*.conf and then it can be overridden again in jail.local. And, any of them can set something that isn't set in any of the rest of them.

So, the resulting config is a combination of all of the files...with local taking precedence over jail.d which takes precedence over jail.conf.

One more note, I'm guessing jail.d is parsed alphabetically (thus the reason for CentOS naming its jail.d file 00-firewalld.conf), with later ones able to override earlier ones. I'll test this to confirm, but I have a strong hunch.

**Comment by jcameron** [comment]
Looks like  currently, webmin only uses entries from the jail.local file that aren't already defined in jail.conf .

**Comment by swelljoe** [comment]
Ah, I wonder if maybe, because this is so complicated (infinite number of locations for a config directive), we should be doing something different here.

Maybe we should get the results of the parsed configs from fail2ban, and then only modify jail.local with the changes we want. e.g. use fail2ban-client to query what the state of things actually is, and then if a user wants to change it, we change it in jail.local. The editor for the other config files can remain, but the intended purpose of jail.local is for ad hoc changes like we're doing...maybe that's where we should be doing them and not worry about the complexity of parsing the thing (and worse, trying to figure out where to add user changes, since the original location of a config option could be in an infinite number of locations).

If we just change jail.local, that'll always apply (it overrides jail.conf and jail.d), and it'll be what the authors of fail2ban intended to happen (I think). We could, instead, make a jail.d/ZZ-webmin.conf, or something, for adding our config options, but I think that's less intuitive.

**Comment by gnadelwartz** [comment]
I second this! Only problem is that all aliases are dereferenced.
we may need both, reading of config files and final config from fail2ban client.

`fail2ban-client -d` ouputs final config in an easy parseable format:

```
['set', 'syslogsocket', 'auto']
['set', 'loglevel', 'INFO']
['set', 'logtarget', '/var/log/fail2ban.log']
['set', 'dbfile', '/var/lib/fail2ban/fail2ban.sqlite3']
['set', 'dbpurgeage', 86400]
['add', 'SSH', 'auto']
['set', 'SSH', 'usedns', 'no']
['set', 'SSH', 'addlogpath', '/var/log/messages', 'head']
['set', 'SSH', 'maxretry', 4]
['set', 'SSH', 'addignoreip', '127.0.0.1']
['set', 'SSH', 'addignoreip', '127.0.0.2']
['set', 'SSH', 'addignoreip', '185.170.115.172']
['set', 'SSH', 'addignoreip', '85.214.248.69']
['set', 'SSH', 'addignoreip', '85.214.248.173']
['set', 'SSH', 'addignoreip', '141.6.0.0/16']
['set', 'SSH', 'addignoreip', '66.249.64.0/19']
['set', 'SSH', 'addignoreip', '207.241.224.0/20']
['set', 'SSH', 'logencoding', 'utf-8']
['set', 'SSH', 'bantime', 3600]
['set', 'SSH', 'ignorecommand', '']
['set', 'SSH', 'findtime', 36000]
['set', 'SSH', 'maxlines', '10']
['set', 'SSH', 'addfailregex', '^\\s*(<[^.]+\\.[^.]+>)?\\s*(?:\\S+ )?(?:kernel: \\[ *\\d+\\.\\d+\\] )?(?:@vserver_\\S+ )
?(?:(?:\\[\\d+\\])?:\\s+[\\[\\(]?sshd(?:\\(\\S+\\))?[\\]\\)]?:?|[\\[\\(]?sshd(?:\\(\\S+\\))?[\\]\\)]?:?(?:\\[\\d+\\])?:?
)?\\s(?:\\[ID \\d+ \\S+\\])?\\s*(?:error: PAM: )?[aA]uthentication (?:failure|error) for .* from <HOST>( via \\S+)?\\s*$
']
.....
['set', 'SSH', 'addjournalmatch', '_SYSTEMD_UNIT=sshd.service', '+', '_COMM=sshd']
['set', 'SSH', 'addaction', 'iptables-allports']
['set', 'SSH', 'action', 'iptables-allports', 'actionban', '<iptables> -I f2b-<name> 1 -s <ip> -j <blocktype>']
['set', 'SSH', 'action', 'iptables-allports', 'actionstop', '<iptables> -D <chain> -p <protocol> -j f2b-<name>\n<iptable
s> -F f2b-<name>\n<iptables> -X f2b-<name>']
['set', 'SSH', 'action', 'iptables-allports', 'actionstart', '<iptables> -N f2b-<name>\n<iptables> -A f2b-<name> -j <ret
urntype>\n<iptables> -I <chain> -p <protocol> -j f2b-<name>']
['set', 'SSH', 'action', 'iptables-allports', 'actionunban', '<iptables> -D f2b-<name> -s <ip> -j <blocktype>']
['set', 'SSH', 'action', 'iptables-allports', 'actioncheck', "<iptables> -n -L <chain> | grep -q 'f2b-<name>[ \\t]'"]
['set', 'SSH', 'action', 'iptables-allports', 'protocol', '"tcp"']
['set', 'SSH', 'action', 'iptables-allports', 'chain', 'INPUT']
['set', 'SSH', 'action', 'iptables-allports', 'lockingopt', '-w']
['set', 'SSH', 'action', 'iptables-allports', 'known/known/name', 'default']
['set', 'SSH', 'action', 'iptables-allports', 'blocktype', 'DROP']
['set', 'SSH', 'action', 'iptables-allports', 'known/lockingopt', '-w']
['set', 'SSH', 'action', 'iptables-allports', 'known/known/port', 'ssh']
['set', 'SSH', 'action', 'iptables-allports', 'known/protocol', 'tcp']
['set', 'SSH', 'action', 'iptables-allports', 'known/known/lockingopt', '-w']
....
['set', 'SSH', 'action', 'sendmail-whois-lines', 'actionstop', '']
['set', 'SSH', 'action', 'sendmail-whois-lines', 'actionstart', '']
['set', 'SSH', 'action', 'sendmail-whois-lines', 'actionunban', '']
['set', 'SSH', 'action', 'sendmail-whois-lines', 'actioncheck', '']
['set', 'SSH', 'action', 'sendmail-whois-lines', 'known/known/sendername', 'Fail2Ban']
['set', 'SSH', 'action', 'sendmail-whois-lines', 'name', 'SSH']
['set', 'SSH', 'action', 'sendmail-whois-lines', 'known/sender', 'fail2ban']
['set', 'SSH', 'action', 'sendmail-whois-lines', 'dest', 'fail2ban@rrr.de']
...
['start', 'SSH']
['start', 'dovecot-pop3imap']
['start', 'VSFTPD']
['start', 'ROUNDCUBE']
['start', 'BadBots']
['start', 'ScanBots']
['start', 'MANYTIMES-24h']
['start', 'BAN-forever']
['start', 'apache-forever']
['start', 'BadBots-forever']
```

**Comment by gnadelwartz** [comment]
what about a different idea for parsing config files: read them in reverse order?
with LIFO there is no need to override directives, because first one found is displayed, older ones are ignored.

**Comment by swelljoe** [comment]
Why do we need to know what file they came from, if we're only changing jail.local? I think that's actually an improvement over trying to keep up with which file it's coming from originally. The docs suggest never changing jail.conf, and there's good arguments against changing existing jail.d files (they're provided by packages from the distro and so could be upgraded in the future).

There's a lot of good reasons to only modify jail.local (except in the plaintext config file editors, which can keep doing what they're doing, but nobody ever messes with those, generally). It removes any doubt about whether it's the option that'll take precedent (local is always parsed last), and it removes any doubt of the OS changing things out from under us (jail.conf and some jail.d files can be modified by the system).

I think we could simplify things a lot, and make them behave actually correctly (right now we have wrong info in the UI), by just taking the parsed output from fail2ban-client, and then only making changes in jail.local.

**Comment by gnadelwartz** [comment]
you are rigth, but the output of fail2ban-client has only the active config, but e.g. not the names/config of inactive jails etc.

so we may need both, fail2ban-client for active config and parsing files for additinal info, e.g. defined but not active jails to have the possibility to activate them in jail.local

**Comment by swelljoe** [comment]
Ah, I didn't think about that. Damn. This remains very complicated.

But, then again, Jamie's already built a parser that works, it's just throwing away some data and not overriding some things that ought to be overridden.

So, module needs to continue parsing the config files, it just needs to be fixed to parse them in the right order, and override things as it finds them. But, there's still no argument against only making changes to jail.local.

**Comment by gnadelwartz** [comment]
We should stay with Jamie's parser, but fix it to read all jail files in correct order.

Additionally it make sense to parse the active config from `fail2ban-client` to have the resulting active config, e.g. running jails, active jail actions and so on. Some minor differences between parsing config and active config are then no more so important.

Changing values only in jail.locals sound good.

BTW: Even if the output of `fail2ban-client -d` look easy parse able in the first place, I have no clue how to translate to an useable (config) array of hashes :-(

**Comment by gnadelwartz** [comment]
I also like the idea of reading the files in rever order to get a virtual LIFO, so the first found config should be the one failtoban is using. Give it a try:

https://github.com/gnadelwartz/webmin/blob/patch-1/fail2ban/fail2ban-lib.pl

https://github.com/gnadelwartz/webmin/commit/ad925746b0cd0227c99cf54849e96ad37994950c

**Comment by chris001** [comment]
Suppose a rule exists in `jail.conf` or `jail.d/*` and is disabled.
Shouldn't the webmin / virtualmin web interface show the user that the rule exists in `jail.conf` and that it's disabled?
When the user wants to enable that rule, shouldn't webmin/virtualmin enable that rule in `jail.local` ?
After enabling the rule in `jail.local`, should the web interface show the rule's enabled in `jail.local` ?

**Comment by gnadelwartz** [comment]
yes yes yes `:-)` should already work this way

**Comment by gnadelwartz** [comment]
@swelljoe  I had a look at the ordre jail files are read, the man page states:

```
jail.d/ and fail2ban.d/

              In addition to .local, for jail.conf or fail2ban.conf file there can be a corresponding .d/ directory containing
              additional .conf files. The order e.g. for jail configuration would be:

              jail.conf
              jail.d/*.conf (in alphabetical order)
              jail.local
              jail.d/*.local (in alphabetical order).
```
so we should make webmin read all jail config files in this order ...

**Comment by gnadelwartz** [comment]
to make things more complex we have also to respect inculde section in every file:

```
       Configuration  files  can include other (defining common variables) configuration files,
       which is often used in Filters and Actions.
       Such inclusions are defined in a section called [INCLUDES]:

       *before* indicates that the specified file is to be parsed before the current file.

       *after*  indicates that the specified file is to be parsed after the current file.

```

**Comment by gnadelwartz** [comment]
from a first view on the source we may change the behaivor of `read_config` file. today read in file.conf and file.local. this has to be changed to read in *all* the realated config files in fail2ban given order.

e.g. if given `&parse_config_file("$config{'config_dir'}/fail2ban.conf");` it should parse all files related to fail2ban conf

1. failban.conf
2. fail2ban.d/*.conf
3. fail2ban.local
4. fail2ban.d/*.conf

**Comment by gnadelwartz** [comment]
seems tom work for me, my jail.local is read as before :-)

![bildschirmfoto vom 2018-01-10 13-55-32](https://user-images.githubusercontent.com/4593242/34774049-3b9af884-f60e-11e7-86ea-6424b507cbf1.png)

@swelljoe can you plaese test it also?

**Comment by gnadelwartz** [comment]
hm, read all files, but does not work as expected ...

**Comment by qooob** [comment]
Why not just create the symlink?

**Comment by gnadelwartz** [comment]
?

**Comment by qooob** [comment]
I mixed up issues.

**Comment by jcameron** [comment]
Would a file in  jail.d ever override something in jail.conf though?

**Comment by gnadelwartz** [comment]
yes,  everything man page and my tests show that everything can be overwritten.

if I e.g define a active jail in jail.conf and redefine as disabled in jail.d its disabled.
if I then enable it in jail.local again its enabled.

**Comment by gnadelwartz** [comment]
OK, I did some more tests on jail.conf, jail.d, jail.local etc.
after each change I've done a `fail2ban-client reload test; fail2ban-client status test`

I created and enabled test jail in jail.conf:
```
[test]
enabled  = true
filter   = vsftpd
logpath  = %(apache_access_log)s
-> jail test active
```

now disable test in jail.d/test.conf
```
[test]
enabled  = false
-> jail test not exist
```

now enable again in jail.local
```
[test]
enabled  = true
-> jail test active
```

now disable in jail.d/notest.local
```
[test]
enabled  = false
-> jail test not exist
```

So YES, files in `jail.d/*.conf` overwrites content of `jail.conf`.
`jail.local` again overrides them again, `jail.d/*.local` overrides them all.

The files in jail.d are read in aphabetical order, but the content is always processed independ of file name: `notest.local` overrides `test.conf`

**Comment by iliajie** [comment]
This issue is fixed by now!
